streaming/base/converters/dataframe_to_mds.py

```diff
@@ -0,0 +1,302 @@
+# Copyright 2023 MosaicML Streaming authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""A utility to convert spark dataframe to MDS."""
+
+import json
+import logging
+import os
+import shutil
+from collections.abc import Iterable
+from typing import Any, Callable, Dict, Iterable, Optional, Tuple, Union
+
+import pandas as pd
+
+from streaming.base.util import get_import_exception_message
+
+try:
+    from pyspark import TaskContext
+    from pyspark.sql.dataframe import DataFrame
+    from pyspark.sql.types import (ArrayType, BinaryType, BooleanType, ByteType, DateType,
+                                   DayTimeIntervalType, DecimalType, DoubleType, FloatType,
+                                   IntegerType, LongType, MapType, ShortType, StringType,
+                                   StructField, StructType, TimestampNTZType, TimestampType)
+except ImportError as e:
+    e.msg = get_import_exception_message(e.name, extra_deps='spark')  # pyright: ignore
+    raise e
+
+from streaming import MDSWriter
+from streaming.base.format.index import get_index_basename
+from streaming.base.format.mds.encodings import _encodings
+from streaming.base.storage.upload import CloudUploader
+
+logger = logging.getLogger(__name__)
+
+
+def is_iterable(obj: object) -> bool:
+    """Check if obj is iterable.
+
+    Args:
+        obj: python object
+    Return:
+        bool: true if obj is iterable false otherwise
+    """
+    return issubclass(type(obj), Iterable)
+
+
+def infer_dataframe_schema(dataframe: DataFrame,
+                           user_defined_cols: Optional[Dict[str, Any]] = None) -> Optional[Dict]:
+    """Retrieve schema to construct a dictionary or do sanity check for MDSWriter.
+
+    Args:
+        dataframe (spark dataframe): dataframe to inspect schema
+        user_defined_cols (Optional[Dict[str, Any]]): user specified schema for MDSWriter
+    Return:
+        If user_defined_cols is None, return schema_dict (dict): column name and dtypes that supported by MDSWriter
+        Else, return None
+    Exceptions:
+        Any of the datatype found to be unsupported by MDSWriter, then raise ValueError
+    """
+    mapping_spark_to_mds = {
```

Comment:
Should we keep this global ? For example, after the import with name as `MAPPING_SPARK_TO_MDS` ?

Reply:
good idea!